File: src-tauri/src/commands/bridge.rs
/// All Tauri invoke commands — the only public interface between UI and Rust.
///
/// Rules:
/// - Every command returns Result<T, String> (Tauri requirement)
/// - State lives in tauri::State wrappers, registered in main.rs
/// - Events are emitted on the AppHandle

use std::sync::{Arc, Mutex};

use tauri::{AppHandle, Manager, State};
use tracing::{error, info};

use crate::{
    audio::{
        devices::{enumerate_audio_devices, AudioDevice},
        virtual_device::{get_virtual_setup, query_virtual_status, VirtualSetupStatus},
        wasapi::ComGuard,
        device_control::{open_control, DeviceKind, VirtualDeviceInfo},
    },
    detection::process::{detect_audio_processes, AudioProcess, ProcessDetector},
    routing::{engine::RoutingEngine, graph::RoutingGraph, node::RouteId},
};

// ── Shared state ──────────────────────────────────────────────────────────

// RoutingEngine is internally synchronized (Arc<Mutex<…>> fields + atomics), so it
// needs no outer Mutex. Sharing it via Arc lets commands run concurrently and lets
// slow operations move onto a blocking thread without holding a global engine lock.
pub struct EngineState(pub Arc<RoutingEngine>);
pub struct DetectorState(pub Mutex<ProcessDetector>);
/// The shared workspace document (scenes) — single source of truth for both UIs
/// (t17 phase B). Same Arc the daemon's /rpc dispatcher uses.
pub struct SceneState(pub Arc<crate::server::scene_store::SceneStore>);
/// Shared application settings (t14) — mirrored + persisted like the scene.
pub struct SettingsState(pub Arc<crate::server::settings_store::SettingsStore>);

// ── Device commands ────────────────────────────────────────────────────────

/// Enumerate audio devices and upgrade render-side virtual endpoints to Virtual.
/// Shared by the Tauri command and the daemon's /rpc dispatcher (t17) so both
/// return identical device lists. Blocking (COM + WASAPI) — call off the async
/// runtime (spawn_blocking) when invoked from async contexts.
///
/// Only render endpoints (Output) become Virtual — capture endpoints (Input) stay
/// Input. CABLE Input is a render endpoint → Virtual (shown as Output target in
/// UI); CABLE Output is a capture endpoint → stays Input (not a render target).
pub fn list_devices_full() -> Result<Vec<AudioDevice>, String> {
    let _com = ComGuard::init().map_err(|e| e.to_string())?;
    let mut devices = enumerate_audio_devices().map_err(|e| e.to_string())?;

    let virtual_status = query_virtual_status(&devices);
    for vd in &virtual_status.devices {
        if let Some(d) = devices.iter_mut().find(|d| d.id == vd.id) {
            if d.device_type == crate::audio::devices::DeviceType::Output {
                d.device_type = crate::audio::devices::DeviceType::Virtual;
            }
        }
    }

    Ok(devices)
}

/// Return all audio devices (input + output + detected virtual).
/// VB-Audio CABLE devices are returned with Nodus-branded names.
#[tauri::command]
pub async fn get_audio_devices(
    _engine: State<'_, EngineState>,
) -> Result<Vec<AudioDevice>, String> {
    list_devices_full()
}

/// Return the current virtual device setup status (NotFound / VbAudio / NodusDriver).
/// UI uses this on startup to decide whether to show the onboarding dialog.
#[tauri::command]
pub async fn get_virtual_setup_status() -> Result<VirtualSetupStatus, String> {
    let _com = ComGuard::init().map_err(|e| e.to_string())?;
    let devices = enumerate_audio_devices().map_err(|e| e.to_string())?;
    Ok(get_virtual_setup(&devices))
}

/// Download and install VB-Audio VBCABLE (Windows only, shows UAC prompt).
/// Poll `get_virtual_setup_status` afterward to confirm success.
#[tauri::command]
pub async fn install_vbcable() -> Result<(), String> {
    crate::audio::virtual_device::setup::install_vbcable().await
}

/// Return whether Windows test signing mode is currently enabled.
/// Used by onboarding UI to show Test Mode option state.
#[tauri::command]
pub async fn is_test_signing_enabled() -> Result<bool, String> {
    Ok(crate::audio::virtual_device::setup::is_test_signing_enabled())
}

// ── Process commands ───────────────────────────────────────────────────────

/// Return currently running audio processes.
#[tauri::command]
pub async fn get_running_audio_processes() -> Result<Vec<AudioProcess>, String> {
    detect_audio_processes().map_err(|e| e.to_string())
}

// ── Routing commands ───────────────────────────────────────────────────────

/// Replace the entire routing graph and restart routing if engine is running.
/// Runs on a blocking thread: the restart settles WASAPI over ~80ms and must not
/// block the async runtime. Per-route volume/mute stay responsive meanwhile.
#[tauri::command]
pub async fn apply_routing_graph(
    graph: RoutingGraph,
    engine: State<'_, EngineState>,
) -> Result<(), String> {
    let engine = Arc::clone(&engine.0);
    tokio::task::spawn_blocking(move || engine.apply_graph(graph))
        .await
        .map_err(|e| e.to_string())?
        .map_err(|e| e.to_string())
}

/// Mute or unmute a specific route (edge) without restarting. Fast, lock-free hot path.
#[tauri::command]
pub async fn set_route_mute(
    route_id: RouteId,
    muted: bool,
    engine: State<'_, EngineState>,
) -> Result<(), String> {
    engine.0.set_route_mute(&route_id, muted).map_err(|e| e.to_string())
}

/// Set volume [0.0 .. 1.0] on a specific route without restarting. Fast, lock-free hot path.
#[tauri::command]
pub async fn set_route_volume(
    route_id: RouteId,
    volume: f32,
    engine: State<'_, EngineState>,
) -> Result<(), String> {
    engine.0.set_route_volume(&route_id, volume).map_err(|e| e.to_string())
}

/// Set stereo balance [-1.0 .. 1.0] on a specific route without restarting.
#[tauri::command]
pub async fn set_route_pan(
    route_id: RouteId,
    pan: f32,
    engine: State<'_, EngineState>,
) -> Result<(), String> {
    engine.0.set_route_pan(&route_id, pan).map_err(|e| e.to_string())
}

/// Start the routing engine (WASAPI setup → blocking thread).
#[tauri::command]
pub async fn start_engine(engine: State<'_, EngineState>) -> Result<(), String> {
    let engine = Arc::clone(&engine.0);
    tokio::task::spawn_blocking(move || engine.start())
        .await
        .map_err(|e| e.to_string())?
        .map_err(|e| e.to_string())
}

/// Whether the routing engine is currently running. Clients read this on mount /
/// reconnect to initialise the Engine button; live changes arrive via the
/// `engine-state` event (t17).
#[tauri::command]
pub fn is_engine_running(engine: State<'_, EngineState>) -> bool {
    engine.0.is_running()
}

/// Stop the routing engine (WASAPI teardown → blocking thread).
#[tauri::command]
pub async fn stop_engine(engine: State<'_, EngineState>) -> Result<(), String> {
    let engine = Arc::clone(&engine.0);
    tokio::task::spawn_blocking(move || engine.stop())
        .await
        .map_err(|e| e.to_string())?
        .map_err(|e| e.to_string())
}

// ── Virtual devices (t5 step 3, S3.5) ────────────────────────────────────────
// Dynamic Nodus virtual devices via the kernel control channel (\\.\NodusControl).
// All open a short-lived handle per call and run on a blocking thread (DeviceIoControl
// is a blocking syscall). No driver / older build → a clear error string.

/// List the driver's device table (2 static + up to 8 dynamic).
#[tauri::command]
pub async fn list_virtual_devices() -> Result<Vec<VirtualDeviceInfo>, String> {
    tokio::task::spawn_blocking(|| {
        let ctl = open_control().map_err(|e| e.to_string())?;
        ctl.list_devices().map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Create a dynamic virtual device (render/capture) with a friendly name; returns it.
#[tauri::command]
pub async fn create_virtual_device(kind: String, name: String) -> Result<VirtualDeviceInfo, String> {
    let k = match kind.as_str() {
        "render" => DeviceKind::Render,
        "capture" => DeviceKind::Capture,
        other => return Err(format!("unknown kind '{other}' (use render|capture)")),
    };
    tokio::task::spawn_blocking(move || {
        let ctl = open_control().map_err(|e| e.to_string())?;
        let id = ctl.create_device(k, None, &name).map_err(|e| e.to_string())?;
        Ok(VirtualDeviceInfo { id, kind: k, name, is_static: false, ring_active: false })
    })
    .await
    .map_err(|e| e.to_string())?
}

/// Destroy a dynamic virtual device by its driver id (1..8; id 0 is refused).
#[tauri::command]
pub async fn remove_virtual_device(id: u32) -> Result<(), String> {
    tokio::task::spawn_blocking(move || {
        let ctl = open_control().map_err(|e| e.to_string())?;
        ctl.destroy_device(id).map_err(|e| e.to_string())
    })
    .await
    .map_err(|e| e.to_string())?
}

// ── Scene sync (t17 phase B) ─────────────────────────────────────────────────

/// Return the current workspace document + revision (single source of truth).
#[tauri::command]
pub fn get_scene(
    scene: State<'_, SceneState>,
) -> crate::server::scene_store::SceneSnapshot {
    scene.0.snapshot()
}

/// Replace the workspace document; persists + broadcasts `scene:snapshot` to the
/// other UIs. `origin` is the caller's client id so it can ignore its own echo.
/// Returns the new revision.
#[tauri::command]
pub fn push_scene(
    doc: serde_json::Value,
    origin: Option<String>,
    scene: State<'_, SceneState>,
) -> u64 {
    scene.0.push(doc, origin)
}

// ── Settings (t14) ───────────────────────────────────────────────────────────

/// Return the current application settings.
#[tauri::command]
pub fn get_settings(
    settings: State<'_, SettingsState>,
) -> crate::server::settings_store::Settings {
    settings.0.get()
}

/// Replace application settings; persists + broadcasts `settings:changed` to the
/// other UIs and applies side effects (e.g. Windows autostart). Returns normalised.
#[tauri::command]
pub fn set_settings(
    next: crate::server::settings_store::Settings,
    settings: State<'_, SettingsState>,
) -> crate::server::settings_store::Settings {
    settings.0.set(next)
}

// ── Background tasks ───────────────────────────────────────────────────────

/// Spawn background tasks that publish events onto the daemon `bus` (t17).
///
/// Single producer → one bus → two consumers: a forwarder in main.rs mirrors the
/// bus to `emit_all` (desktop webview) while each WS connection mirrors it to its
/// socket (Web-UI / Claude-preview). This same path also carries `scene:snapshot`
/// (Phase B), so a scene change from a web client reaches the desktop too.
pub fn setup_background_tasks(
    handle: AppHandle,
    bus: crate::server::EventBus,
    settings: Arc<crate::server::settings_store::SettingsStore>,
) {
    // Process detector — publishes "process-changed" when the list changes. Interval
    // comes from settings (applied at launch). The ProcessDetector's background thread
    // holds Arc refs, so dropping the local detector here is fine.
    {
        let bus_proc = bus.clone();
        let detector = ProcessDetector::new();
        detector.start(settings.get().scan_interval(), move |procs| {
            match serde_json::to_value(&procs) {
                Ok(payload) => {
                    let _ = bus_proc.send(crate::server::ServerEvent {
                        event: "process-changed".into(),
                        payload,
                    });
                }
                Err(e) => error!("failed to serialize process list: {e}"),
            }
        });
        // detector drops here; background thread is kept alive by its own Arc refs
    }

    // Publish the initial device list on a background thread.
    let bus_dev = bus.clone();
    std::thread::spawn(move || {
        // Use the shared helper so the startup event carries the same Virtual
        // upgrades as get_audio_devices / the /rpc dispatcher.
        match list_devices_full() {
            Ok(devices) => {
                if let Ok(payload) = serde_json::to_value(&devices) {
                    let _ = bus_dev.send(crate::server::ServerEvent {
                        event: "audio-devices-changed".into(),
                        payload,
                    });
                }
            }
            Err(e) => error!("failed to enumerate devices on startup: {e}"),
        }
    });

    // VU meter — publishes "volume-levels" at ~15fps when the engine is running, and
    // only when the levels actually changed: every event triggers a WebView repaint,
    // which is expensive on weak GPUs (Pentium N4200 field test: WebView2 GPU
    // process at ~27% CPU). Payload: {device_id: level_0_to_1}.
    let handle_levels = handle.clone();
    let bus_levels = bus.clone();
    let settings_levels = settings.clone();
    std::thread::spawn(move || {
        let mut prev: std::collections::HashMap<String, f32> = Default::default();
        let mut was_running = false;
        let publish = |payload: serde_json::Value, bus: &crate::server::EventBus| {
            let _ = bus.send(crate::server::ServerEvent {
                event: "volume-levels".into(),
                payload,
            });
        };
        loop {
            // VU refresh rate is configurable live (t14): re-read each iteration.
            let cfg = settings_levels.get();
            std::thread::sleep(cfg.vu_interval());
            let engine = handle_levels.state::<EngineState>();
            // Engine is lock-free to query now; get_levels takes only the
            // short-lived internal captures/routes locks.
            let running = engine.0.is_running();
            // Broadcast engine on/off transitions so EVERY client's Engine button
            // stays in sync — the engine is a single shared instance, so its state
            // is the source of truth, not any one UI's local flag (t17).
            if running != was_running {
                was_running = running;
                let _ = bus_levels.send(crate::server::ServerEvent {
                    event: "engine-state".into(),
                    payload: serde_json::json!(running),
                });
            }
            // VU disabled (t14) or engine stopped → make sure meters are zeroed once,
            // then idle. Engine-state above still flows so the button stays in sync.
            if !running || !cfg.vu_enabled {
                if !prev.is_empty() {
                    prev.clear();
                    if let Ok(payload) = serde_json::to_value(&prev) {
                        publish(payload, &bus_levels);
                    }
                }
                continue;
            }
            let levels = engine.0.get_levels();
            let changed = levels.len() != prev.len()
                || levels
                    .iter()
                    .any(|(k, v)| (prev.get(k).copied().unwrap_or(-1.0) - v).abs() > 0.01);
            if changed {
                if let Ok(payload) = serde_json::to_value(&levels) {
                    publish(payload, &bus_levels);
                }
                prev = levels;
            }
        }
    });

    info!("background tasks started");
}

// ── Serialization tests ────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use crate::{
        audio::devices::{AudioDevice, DeviceType},
        detection::process::{AudioProcess, SourceType},
        routing::{
            graph::RoutingGraph,
            node::{Node, NodeType, Route},
        },
    };

    #[test]
    fn audio_device_serializes() {
        let d = AudioDevice {
            id: "test-id".into(),
            name: "Test Device".into(),
            device_type: DeviceType::Output,
            is_default: true,
            original_name: None,
            is_virtual: false,
        };
        let json = serde_json::to_string(&d).unwrap();
        let back: AudioDevice = serde_json::from_str(&json).unwrap();
        assert_eq!(back.id, d.id);
        assert_eq!(back.device_type, DeviceType::Output);
    }

    #[test]
    fn audio_process_serializes() {
        let p = AudioProcess {
            exe_name: "arma3_x64.exe".into(),
            pid: 1234,
            display_name: "Arma 3".into(),
            source_type: SourceType::Game,
            icon: None,
        };
        let json = serde_json::to_string(&p).unwrap();
        assert!(json.contains("arma3_x64.exe"));
        assert!(json.contains("\"game\""));
    }

    #[test]
    fn routing_graph_round_trips() {
        let src = Node::new(NodeType::Source, "Arma 3", "dev-arma");
        let dst = Node::new(NodeType::Output, "Headphones", "dev-hp");
        let route = Route::new(src.id.clone(), dst.id.clone());
        let graph = RoutingGraph {
            nodes: vec![src, dst],
            routes: vec![route],
        };

        let json = serde_json::to_string(&graph).unwrap();
        let back: RoutingGraph = serde_json::from_str(&json).unwrap();
        assert_eq!(back.nodes.len(), 2);
        assert_eq!(back.routes.len(), 1);
    }
}

File: src/bridge.ts
/**
 * bridge.ts — typed Tauri bridge (redesign R3).
 *
 * Wraps the Rust `invoke` commands and events with the exact types from the
 * backend (commands/bridge.rs, audio/devices.rs, detection/process.rs,
 * routing/node.rs). Falls back gracefully when there is no Tauri runtime
 * (plain Vite preview): commands resolve to null/empty and listeners are no-ops,
 * so the UI runs in the browser with sample data.
 *
 * This file is the single contract between the React UI and the Rust engine —
 * changing a command/event here must match the Rust side.
 */

// ── Backend types (mirror the Rust serde shapes) ────────────────────────────

export type DeviceType = 'input' | 'output' | 'virtual';

export interface AudioDevice {
  id: string;
  name: string;
  /** Data flow: input = capture, output = render (drives port direction). */
  device_type: DeviceType;
  is_default: boolean;
  original_name?: string | null;
  /** Software/virtual device (VB-Cable, VoiceMeeter, MIXLINE, Nodus…) — t9. */
  is_virtual?: boolean;
}

export type SourceType =
  | 'game'
  | 'chat'
  | 'voice'
  | 'music'
  | 'browser'
  | 'recording'
  | 'system'
  | 'unknown';

export interface AudioProcess {
  exe_name: string;
  pid: number;
  display_name: string;
  source_type: SourceType;
  /** App icon extracted from the .exe as a PNG data URL (R7). null if unavailable. */
  icon?: string | null;
}

export type BackendNodeType = 'source' | 'output' | 'splitter' | 'mixer' | 'virtual';

export interface BackendNode {
  id: string;
  node_type: BackendNodeType;
  label: string;
  device_id: string;
  exe_name?: string | null;
}

export interface BackendRoute {
  id: string;
  from_node: string;
  to_node: string;
  volume: number; // 0..1
  muted: boolean;
  pan: number; // -1..1
}

export interface RoutingGraph {
  nodes: BackendNode[];
  routes: BackendRoute[];
}

export interface VirtualSetupStatus {
  kind: 'not_found' | 'vb_audio' | 'nodus_driver';
  endpoints: string[];
  message: string;
}

/** volume-levels payload: keyed by device id OR exe name → level 0..1. */
export type VolumeLevels = Record<string, number>;

export type NodusEvent =
  | 'audio-devices-changed'
  | 'process-changed'
  | 'volume-levels'
  | 'engine-state';

// ── Runtime detection + lazy Tauri API ──────────────────────────────────────

export const isTauri =
  typeof window !== 'undefined' && '__TAURI_IPC__' in (window as unknown as Record<string, unknown>);

// ── Web daemon transport (t17 phase A) ──────────────────────────────────────
// When NOT running inside Tauri, the UI can still drive a real engine by talking
// to the embedded Nodus daemon over HTTP (/rpc) + WebSocket (/ws). The daemon
// address is taken from (in order) the URL `?daemon=` param, the VITE_NODUS_DAEMON
// env var, or localStorage 'nodus.daemon'. The token comes from `?token=`,
// VITE_NODUS_TOKEN, or localStorage 'nodus.token'. Accepts `host:port`,
// `http://host:port`, or `ws://host:port`. No daemon configured → the old browser
// behaviour (commands resolve to null, listeners are no-ops, sample data).

interface DaemonCfg {
  http: string;
  ws: string;
  token: string;
}

function readDaemonCfg(): DaemonCfg | null {
  if (typeof window === 'undefined') return null;
  const params = new URLSearchParams(window.location.search);
  const env = (import.meta as unknown as { env?: Record<string, string> }).env ?? {};
  const ls = (key: string) => {
    try {
      return typeof localStorage !== 'undefined' ? localStorage.getItem(key) : null;
    } catch {
      return null;
    }
  };
  const raw = params.get('daemon') || env.VITE_NODUS_DAEMON || ls('nodus.daemon') || '';
  if (!raw) return null;
  const secure = /^(https|wss):/i.test(raw);
  const host = raw
    .replace(/^wss?:\/\//i, '')
    .replace(/^https?:\/\//i, '')
    .replace(/\/+$/, '');
  const token = params.get('token') || env.VITE_NODUS_TOKEN || ls('nodus.token') || '';
  return {
    http: `${secure ? 'https' : 'http'}://${host}`,
    ws: `${secure ? 'wss' : 'ws'}://${host}`,
    token,
  };
}

const _daemon: DaemonCfg | null = !isTauri ? readDaemonCfg() : null;
/** True when this browser session is wired to a live Nodus daemon (t17). */
export const isDaemon = _daemon !== null;

type InvokeFn = (cmd: string, args?: Record<string, unknown>) => Promise<unknown>;
type ListenFn = (
  event: string,
  handler: (e: { payload: unknown }) => void,
) => Promise<() => void>;

let _invoke: InvokeFn | null = null;
let _listen: ListenFn | null = null;

async function getInvoke(): Promise<InvokeFn | null> {
  if (!isTauri) return null;
  if (!_invoke) {
    const mod = await import('@tauri-apps/api/tauri');
    _invoke = mod.invoke as InvokeFn;
  }
  return _invoke;
}

async function getListen(): Promise<ListenFn | null> {
  if (!isTauri) return null;
  if (!_listen) {
    const mod = await import('@tauri-apps/api/event');
    _listen = mod.listen as unknown as ListenFn;
  }
  return _listen;
}

async function call<T>(cmd: string, args?: Record<string, unknown>): Promise<T | null> {
  if (isTauri) {
    const fn = await getInvoke();
    if (!fn) return null;
    return (await fn(cmd, args)) as T;
  }
  if (_daemon) {
    const url = `${_daemon.http}/rpc${_daemon.token ? `?token=${encodeURIComponent(_daemon.token)}` : ''}`;
    const res = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ cmd, args: args ?? {} }),
    });
    if (!res.ok) throw new Error(`rpc ${cmd}: HTTP ${res.status}`);
    const data = (await res.json()) as { ok?: boolean; result?: unknown; error?: string };
    if (data.ok === false) throw new Error(data.error || `rpc ${cmd} failed`);
    return (data.result ?? null) as T;
  }
  return null;
}

// ── Daemon WebSocket: a single shared connection fanned out to all listeners ──

type WsHandler = (payload: unknown) => void;
const _wsHandlers = new Map<string, Set<WsHandler>>();
let _ws: WebSocket | null = null;
let _wsRetry: ReturnType<typeof setTimeout> | null = null;

function ensureWs(): void {
  if (!_daemon || _ws) return;
  const url = `${_daemon.ws}/ws${_daemon.token ? `?token=${encodeURIComponent(_daemon.token)}` : ''}`;
  try {
    const ws = new WebSocket(url);
    _ws = ws;
    ws.onmessage = (ev) => {
      try {
        const { event, payload } = JSON.parse(ev.data as string) as {
          event: string;
          payload: unknown;
        };
        _wsHandlers.get(event)?.forEach((h) => h(payload));
      } catch {
        /* ignore malformed frame */
      }
    };
    ws.onclose = () => {
      _ws = null;
      if (_wsHandlers.size) scheduleWsRetry();
    };
    ws.onerror = () => {
      try {
        ws.close();
      } catch {
        /* noop */
      }
    };
  } catch {
    _ws = null;
    scheduleWsRetry();
  }
}

function scheduleWsRetry(): void {
  if (_wsRetry) return;
  _wsRetry = setTimeout(() => {
    _wsRetry = null;
    ensureWs();
  }, 1500);
}

function daemonListen(event: string, handler: WsHandler): () => void {
  let set = _wsHandlers.get(event);
  if (!set) {
    set = new Set();
    _wsHandlers.set(event, set);
  }
  set.add(handler);
  ensureWs();
  return () => {
    const s = _wsHandlers.get(event);
    if (!s) return;
    s.delete(handler);
    if (!s.size) _wsHandlers.delete(event);
  };
}

// ── Generic event channel (used by the quick-controls flyout ↔ main, t13) ────

type EmitFn = (event: string, payload?: unknown) => Promise<void>;
let _emit: EmitFn | null = null;

/** Emit a Tauri event to all windows. No-op in the browser. */
export async function emitEvent(event: string, payload?: unknown): Promise<void> {
  if (!isTauri) return;
  if (!_emit) {
    const mod = await import('@tauri-apps/api/event');
    _emit = mod.emit as unknown as EmitFn;
  }
  await _emit(event, payload);
}

/** Listen to any event by name. Returns an unsubscribe fn (no-op when offline). */
export async function listenAny<T>(
  event: string,
  handler: (payload: T) => void,
): Promise<() => void> {
  if (isTauri) {
    const listen = await getListen();
    if (!listen) return () => {};
    return listen(event, (e) => handler(e.payload as T));
  }
  if (_daemon) return daemonListen(event, (p) => handler(p as T));
  return () => {};
}

/** Which Tauri window this document is (by ?w= query). 'quick' = the flyout. */
export function windowKind(): 'main' | 'quick' {
  if (typeof window === 'undefined') return 'main';
  return new URLSearchParams(window.location.search).get('w') === 'quick' ? 'quick' : 'main';
}

// ── Commands ────────────────────────────────────────────────────────────────

export async function getAudioDevices(): Promise<AudioDevice[]> {
  return (await call<AudioDevice[]>('get_audio_devices')) ?? [];
}

export async function getRunningAudioProcesses(): Promise<AudioProcess[]> {
  return (await call<AudioProcess[]>('get_running_audio_processes')) ?? [];
}

export function applyRoutingGraph(graph: RoutingGraph): Promise<unknown> {
  return call('apply_routing_graph', { graph });
}

// Tauri 1.x maps JS camelCase invoke args to snake_case Rust params, so the key
// MUST be `routeId` (not `route_id`) or the command's `route_id` arg is missing
// and the invoke is silently rejected. We send both keys to be safe. (t16 fix)
export function setRouteMute(routeId: string, muted: boolean): Promise<unknown> {
  return call('set_route_mute', { routeId, route_id: routeId, muted });
}

export function setRouteVolume(routeId: string, volume: number): Promise<unknown> {
  return call('set_route_volume', { routeId, route_id: routeId, volume });
}

export function setRoutePan(routeId: string, pan: number): Promise<unknown> {
  return call('set_route_pan', { routeId, route_id: routeId, pan });
}

export function startEngine(): Promise<unknown> {
  return call('start_engine');
}

export function stopEngine(): Promise<unknown> {
  return call('stop_engine');
}

/** Whether the shared engine is currently running (read on mount; live updates
 *  arrive via the `engine-state` event). False when there is no live backend. */
export async function isEngineRunning(): Promise<boolean> {
  return (await call<boolean>('is_engine_running')) ?? false;
}

export async function getVirtualSetupStatus(): Promise<VirtualSetupStatus> {
  return (
    (await call<VirtualSetupStatus>('get_virtual_setup_status')) ?? {
      kind: 'not_found',
      endpoints: [],
      message: '',
    }
  );
}

export function installVbcable(): Promise<unknown> {
  return call('install_vbcable');
}

export async function isTestSigningEnabled(): Promise<boolean> {
  return (await call<boolean>('is_test_signing_enabled')) ?? false;
}

/** Pin/unpin the tray flyout (when pinned it stays open on focus loss). */
export function setFlyoutPinned(pinned: boolean): Promise<unknown> {
  return call('set_flyout_pinned', { pinned });
}

/** Reveal the main window from Rust (the flyout's "Dashboard") — works even when
 *  the main window is hidden to the tray (its own JS is throttled while hidden). */
export function showMainWindow(): Promise<unknown> {
  return call('show_main_window');
}

/** Daemon URL + token to open the Web-UI / hand to Claude-preview (t17). Tauri only. */
export interface ServerInfo {
  url: string;
  token: string;
}
export async function getServerInfo(): Promise<ServerInfo | null> {
  if (!isTauri) return null;
  return call<ServerInfo>('get_server_info');
}

// ── Scene sync (t17 phase B) ─────────────────────────────────────────────────
// The workspace document `{ tabs, activeId }` lives in the daemon as the single
// source of truth. A client pushes the whole document after a local mutation;
// the daemon persists + broadcasts `scene:snapshot` so the other UI mirrors it.

/** The daemon's workspace document + monotonic revision (+ origin on broadcasts). */
export interface SceneSnapshot {
  doc: unknown;
  rev: number;
  origin?: string | null;
}

/** Pull the current workspace document (null when there is no live backend). */
export async function getScene(): Promise<SceneSnapshot | null> {
  if (!isTauri && !_daemon) return null;
  return call<SceneSnapshot>('get_scene');
}

/** Push the whole workspace document; returns the new revision (or null offline). */
export async function pushScene(doc: unknown, origin: string): Promise<number | null> {
  if (!isTauri && !_daemon) return null;
  const r = await call<{ rev: number } | number>('push_scene', { doc, origin });
  if (r == null) return null;
  return typeof r === 'number' ? r : r.rev;
}

// ── Settings (t14) ────────────────────────────────────────────────────────────
// Mirrored + persisted in the daemon like the scene; field names match the Rust
// serde shape (snake_case). Broadcast as `settings:changed`.

export interface Settings {
  // performance (live)
  vu_enabled: boolean;
  vu_fps: number;
  process_scan_secs: number;
  // server / browser access (applied at next launch)
  server_port: number;
  server_lan: boolean;
  // app behavior
  start_engine_on_launch: boolean;
  close_to_tray: boolean;
  start_with_windows: boolean;
}

/** Defaults mirroring Rust `Settings::default()` — used before hydrate / offline. */
export const DEFAULT_SETTINGS: Settings = {
  vu_enabled: true,
  vu_fps: 15,
  process_scan_secs: 2,
  server_port: 7878,
  server_lan: false,
  start_engine_on_launch: false,
  close_to_tray: true,
  start_with_windows: false,
};

export async function getSettings(): Promise<Settings | null> {
  if (!isTauri && !_daemon) return null;
  return call<Settings>('get_settings');
}

/** Replace settings; persists + broadcasts + applies side effects. Returns normalised. */
export async function setSettings(next: Settings): Promise<Settings | null> {
  if (!isTauri && !_daemon) return null;
  return call<Settings>('set_settings', { next });
}

// ── Window controls (custom title bar) ──────────────────────────────────────
// Lazy import of @tauri-apps/api/window so the plain Vite preview (no Tauri)
// never loads it; every control is a no-op in the browser.

type AppWindow = {
  minimize: () => Promise<void>;
  toggleMaximize: () => Promise<void>;
  close: () => Promise<void>;
  hide: () => Promise<void>;
  show: () => Promise<void>;
  unminimize: () => Promise<void>;
  setFocus: () => Promise<void>;
  isMaximized: () => Promise<boolean>;
  onResized: (cb: () => void) => Promise<() => void>;
};

let _appWindow: AppWindow | null = null;

async function getWindow(): Promise<AppWindow | null> {
  if (!isTauri) return null;
  if (!_appWindow) {
    const mod = await import('@tauri-apps/api/window');
    _appWindow = mod.appWindow as unknown as AppWindow;
  }
  return _appWindow;
}

export async function winMinimize(): Promise<void> {
  await (await getWindow())?.minimize();
}
export async function winToggleMaximize(): Promise<void> {
  await (await getWindow())?.toggleMaximize();
}
export async function winClose(): Promise<void> {
  await (await getWindow())?.close();
}
/** Hide this window (used by the tray flyout's close — never destroys it). */
export async function winHide(): Promise<void> {
  await (await getWindow())?.hide();
}
/** Reveal this window from any state (hidden to tray OR minimized to taskbar)
 *  and bring it to front — the flyout's "Dashboard" must work every time. */
export async function winShow(): Promise<void> {
  const w = await getWindow();
  if (!w) return;
  // Independent + order-safe: show() un-hides from the tray; unminimize()
  // restores from the taskbar; setFocus() raises. Each is guarded so one
  // failing (e.g. unminimize on a non-minimized window) can't abort the rest.
  await w.show().catch(() => {});
  await w.unminimize().catch(() => {});
  await w.setFocus().catch(() => {});
}
export async function winIsMaximized(): Promise<boolean> {
  return (await (await getWindow())?.isMaximized()) ?? false;
}
/** Subscribe to window resize (to refresh the maximize/restore icon). No-op in browser. */
export async function onWinResize(cb: () => void): Promise<() => void> {
  const w = await getWindow();
  if (!w) return () => {};
  return w.onResized(cb);
}

// ── Events ──────────────────────────────────────────────────────────────────

/** Subscribe to a backend event. Returns an unsubscribe fn (no-op when offline). */
export async function listenToEvent<T>(
  event: NodusEvent,
  handler: (payload: T) => void,
): Promise<() => void> {
  if (isTauri) {
    const listen = await getListen();
    if (!listen) return () => {};
    return listen(event, (e) => handler(e.payload as T));
  }
  if (_daemon) return daemonListen(event, (p) => handler(p as T));
  return () => {};
}
